src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs

```diff
@@ -421,5 +434,27 @@ public TrialResult Run(TrialSettings settings)
 
             throw new ArgumentException($"The runner metric manager is of type {_metricManager.GetType()} which expected to be of type {typeof(ITrainTestDatasetManager)} or {typeof(ICrossValidateDatasetManager)}");
         }
+
+        public Task<TrialResult> RunAsync(TrialSettings settings, CancellationToken ct)
```

Comment:
>RunAsync [](http://example.com/codeflow?start=33&length=8)

I don't know if this is a pattern used in other places but I have some hard time to understand how this is really execute as async? 

`Task.FromResult` will always return a completed task. From the code, `Run(settings)` will always run synchronously and not async. Also, any user calling `await RunAsync(...)`  will always not wait at al as the returned task will be completed. 

am I missing anything here?

Reply:
`RunAsync` allows `AutoMLExperiment` to pass a cancellation token and cancel running trials if necessary, while `Run` doesn't.

Reply:
Mostly because of cancellation pattern. `AutoMLExperiment` provides an async api to start training which implements cancellation pattern. During automl training, what `AutoMLExperiment` does is basically kick off trials in a for-loop and check if cancellation token get triggered and that's how cancellation pattern being implemented. So it would be nice if trial runner also supports the cancellation pattern, which accepts a cancellation token and return a Task.

So to summarize,  `ITrialRunner.RunAsync` returning a task is more required by `AutoMLExperiment` class which makes cancellation easier to be done. It's just ML.Net doesn't support async training so we need to wrap it in a task.

Reply:
Can we make `ITrialRunner.RunAsync` return non completed task and let the caller await for this task? or can this cause a problem? I mean let `Run(settings)` runs inside a task which we return from `RunAsync` and somehow make this task support cancellation. Or we shouldn't claim `RunAsync` is async method and just rename it to something like `RunWithCancellationToken` and not having it return a task at all.

Reply:
Oh sure